controllers/provisioning_controller.go

```diff
@@ -298,3 +282,57 @@ func removeString(slice []string, s string) (result []string) {
 	}
 	return
 }
+
+func (r *ProvisioningReconciler) waitForDeploymentRollout(deployment *appsv1.Deployment, pollInterval, rolloutTimeout time.Duration) error {
+	var lastError error
+	err := wait.Poll(pollInterval, rolloutTimeout, func() (bool, error) {
```

Comment:
IIUC the `wait.Poll` can block, in the worst case, the reconcile loop up to 5 minutes. In such case I'd suggest to adopt a non-blocking design for the reconcile loop, such as tracking the state accordingly and rescheduling the request, so that the operator could continue to work on a regular basis.

Reply:
It looks like this implementation is lifted from the machine-api-operator. @sadasu did you end up changing any of it?

I agree that normally we would want the operator to support more parallelism, but we expect our API resource to be a singleton in this case, right? So maybe blocking is OK in this case, especially if it simplifies the operator logic.

It would be nice if we could put this function in another package, but we can clean that up in a separate PR.

Reply:
Yes, re-using MAO logic here. That is our baseline and we can improve upon it based on our current discussion. Having the wait.Poll() allows us to put the CO is Available state right after the metal3 pod becomes available. It is also really useful to figure out when to put the CO in degraded state. (We don't want to do that after the first failure.)

Yes, I could move it to another package.

If we were to remove the wait, we could maintain a deploymentElapsedCount/Time value and keep decrementing it every time the metal3 deployment is not available in each reconcile loop. Once deploymentElapsedCount/Time hits zero we mark CO as Degraded and success at any point would put the CI in Available state. Is that an alternate approach that seems reasonable?

Reply:
personally i think you should ditch the Poll and just do a once off check. if it is not "up" just return
```
return ctrl.Result{RequeueAfter: time.Minute}, nil
```
Then just let it run the reconcile again

Reply:
In order to tell when to mark the CO as failing, it does seem like we need to know when we started to try to launch/update the deployment. Does the deployment (or pod?) resource provide enough status information for us to compute that time just by looking at it, or do we really have to track it ourselves? It would be nice if we could avoid reverse-engineering failure detection that the deployment controller should be able to give us.

Reply:
Updated logic around how we determine the state of the deployment rollout and then set the CO state accordingly. Current approach might still need some tweaking but looking for feedback regarding general direction.